packages/devtools/lib/src/timeline/timeline_protocol.dart

```diff
@@ -580,67 +582,71 @@ class TimelineFrame {
+  /// use this value to determine whether a TimelineEvent fits within the
+  /// frame's time boundaries.
+  int get pipelineItemStartTime => _pipelineItemStartTime;
+  int _pipelineItemStartTime;
+  set pipelineItemStartTime(int time) =>
+      _pipelineItemStartTime = nullSafeMin(_pipelineItemStartTime, time);
+
+  /// Pipeline item end time in micros.
   ///
-  /// We take the max of [gpuEndTime] and [_endTime] because we use an epsilon
-  /// when determining if an event fits within frame boundaries. Therefore,
-  /// there is a chance that [gpuEndTime] could be greater than [_endTime].
-  int get endTime => nullSafeMax(_endTime, gpuEndTime);
-  int _endTime;
-  set endTime(int time) => _endTime = nullSafeMax(_endTime, time);
+  /// This stores the timestamp of the end pipeline event for this frame. We use
+  /// this value to determine whether a TimelineEvent fits within the frame's
+  /// time boundaries.
+  int get pipelineItemEndTime => _pipelineItemEndTime;
+  int _pipelineItemEndTime;
+  set pipelineItemEndTime(int time) =>
+      _pipelineItemEndTime = nullSafeMax(_pipelineItemEndTime, time);
+
+  bool get isWellFormed =>
+      _pipelineItemStartTime != null && _pipelineItemEndTime != null;
 
-  bool get isWellFormed => _startTime != null && _endTime != null;
+  /// Frame start time in micros.
+  int get startTime => _uiStartTime;
+
+  /// Frame end time in micros.
+  int get endTime => _gpuEndTime;
 
   /// Duration the frame took to render in micros.
   int get duration =>
       endTime != null && startTime != null ? endTime - startTime : null;
 
   // Timing info for UI portion of the frame.
-  int get uiStartTime => uiEventFlow?.startTime;
+  int get _uiStartTime => uiEventFlow?.startTime;
 
-  int get uiEndTime => uiStartTime != null ? uiStartTime + uiDuration : null;
+  int get _uiEndTime => _uiStartTime != null ? _uiStartTime + uiDuration : null;
 
   int get uiDuration => uiEventFlow?.duration;
 
   double get uiDurationMs => uiDuration != null ? uiDuration / 1000 : null;
 
   // Timing info for GPU portion of the frame.
-  int get gpuStartTime => gpuEventFlow?.startTime;
+  int get _gpuStartTime => gpuEventFlow?.startTime;
 
-  int get gpuEndTime =>
-      gpuStartTime != null ? gpuStartTime + gpuDuration : null;
+  int get _gpuEndTime =>
+      _gpuStartTime != null ? _gpuStartTime + gpuDuration : null;
 
   int get gpuDuration => gpuEventFlow?.duration;
 
   double get gpuDurationMs => gpuDuration != null ? gpuDuration / 1000 : null;
 
-  bool get isUiSlow => uiDurationMs > targetMaxDuration / 2;
-
-  bool get isGpuSlow => gpuDurationMs > targetMaxDuration / 2;
-
   @override
   String toString() {
     return 'Frame $id - [start: $startTime], [end: $endTime],'
-        'ui: [start $uiStartTime end $uiEndTime], gpu: [start: $gpuStartTime'
-        ' end $gpuEndTime]';
+        'ui: [start $_uiStartTime end $_uiEndTime], gpu: [start: $_gpuStartTime'
```

Comment:
having a TimeRange class with a nice toString will also simplify this toString and fix inconsistencies in it.

Reply:
done.